### PR TITLE
Catalog app deployment

### PR DESCRIPTION
**Pull request by tomahawk28**
Creating the appDeployment CRD after the app catalog story
Issue linked: giantswarm/giantswarm#4845
previous PR: https://github.com/giantswarm/apiextensions/pull/178

### CODE REVIEW

File: pkg/apis/app/v1alpha1/app_deployment_types.go [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata"`
+	Spec              AppDeploymentSpec `json:"spec"`
+}
+
+type AppDeploymentSpec struct {
+	// Catalog is the name of the app deployment for this CR
+	// e.g. giant-swarm
+	Catalog string `json:"catalog" yaml:"catalog"`
+	App     string `json:"app" yaml:"app"`
```
Comment (tomahawk28) [inline pkg/apis/app/v1alpha1/app_deployment_types.go]: It's not clear to me what is the actual difference between catalog and app in here. Any suggestion?

File: pkg/apis/app/v1alpha1/app_deployment_types.go [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+	// Release is the version of this app which we would like to use.
+	Release string `json:"release" yaml:"release"`
+	// KubeContext is the tenant cluster-based context name which point to specific kubeConfig as well.
+	KubeContext string `json:"kubeContext" yaml:"kubeContext"`
+	// Namespace is the tenant cluster-based namespace where this app would be eventually located.
+	Namespace string                  `json:"namespace" yaml:"namespace"`
```
Comment (tomahawk28) [inline pkg/apis/app/v1alpha1/app_deployment_types.go]: If `kubeContext` is here, do we really need `namespace`? I looked upon the relevant document and it saying that `namespace` is part of kube context. https://kubernetes.io/docs/tasks/access-application-cluster/configure-access-multiple-clusters/#define-clusters-users-and-contexts

File: pkg/apis/app/v1alpha1/app_deployment_types.go [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+// NewAppDeploymentCRD returns a new custom resource definition for AppDeployment.
+// This might look something like the following.
+//
+//     apiVersion: apiextensions.k8s.io/v1beta1
+//     kind: CustomResourceDefinition
+//     metadata:
+//       name: appdeployment.app.giantswarm.io
+//     spec:
+//       group: app.giantswarm.io
+//       scope: Cluster
```
Comment (rossf7) [inline pkg/apis/app/v1alpha1/app_deployment_types.go]: This should also be `Namespaced`.

File: pkg/apis/app/v1alpha1/app_deployment_types.go [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+//       version: v1alpha1
+//       names:
+//         kind: AppDeployment
+//         plural: appdeployments
+//         singular: appdeployments
```
Comment (rossf7) [inline pkg/apis/app/v1alpha1/app_deployment_types.go]: This should also be `appdeployment`.

File: pkg/apis/app/v1alpha1/app_deployment_types.go (hunk 1 above)
Comment (rossf7) [reply pkg/apis/app/v1alpha1/app_deployment_types.go]: Lets use `prometheus` here as this is the first app we will be making available.

File: pkg/apis/app/v1alpha1/app_deployment_types.go (hunk 2 above)
Comment (rossf7) [reply pkg/apis/app/v1alpha1/app_deployment_types.go]: Yes they are separate. The `kubeContext` will have the namespace where the operator should connect to deploy the app, e.g `giantswarm` to connect to our Tiller pod.

This namespace is where the app should be deployed, e.g `giantswarm-prometheus`.

**Review by rossf7 — COMMENTED** [review]
Direction is good. Godoc needs updated and answered your questions.

An admin point. We don't use the `help-wanted` label internally. You can ask for help in any PR! Just ping the relevant people or teams in the PR.

File: pkg/apis/app/v1alpha1/app_deployment_types.go (hunk 1 above)
Comment (rossf7) [reply pkg/apis/app/v1alpha1/app_deployment_types.go]: We need godoc and examples for each field. The examples are to make the CRD easier to read.

Lets align with the example in the spec PR.

```
spec:
  catalog: "giant-swarm"
  app: “kubernetes-prometheus”
  release: 1.0.0
  kubeContext: “giantswarm-12345”
  namespace: “monitoring”
```

**Review by rossf7 — COMMENTED** [review]
Direction is good here. Lets get the examples sorted and then it can be submitted for review.

**Comment by xh3b4sd** [comment]
FYI the `apiextensions` are more and more becoming a directly user facing API. So @giantswarm/sig-ux is a little bit concerned with that. Let's always request their feedback on PRs like this here. To make it easier to reason about the implications and how it looks like we would like to have an example YAML of the added CRs in the godoc. Can you please add these? That would be super cool.

File: pkg/apis/app/v1alpha1/app_deployment_types.go [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+func NewAppDeploymentCRD() *apiextensionsv1beta1.CustomResourceDefinition {
+	return &apiextensionsv1beta1.CustomResourceDefinition{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: apiextensionsv1beta1.SchemeGroupVersion.String(),
+			Kind:       "CustomResourceDefinition",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "appdeployments.app.giantswarm.io",
+		},
+		Spec: apiextensionsv1beta1.CustomResourceDefinitionSpec{
```
Comment (xh3b4sd) [inline pkg/apis/app/v1alpha1/app_deployment_types.go]: Can you please add the status subresources to the CRD? We should always add these by default to all new types.

**Review by xh3b4sd — DISMISSED** [review]
Note that in some places we started to add another helper function for the type meta. When you think this is useful you can also add it. See https://github.com/giantswarm/apiextensions/blob/858ead97a5762244870ebbe3a83503f5132b2fde/pkg/apis/core/v1alpha1/drainer_types.go#L67-L72 for instance.

**Review by rossf7 — CHANGES_REQUESTED** [review]
Godoc isn't accurate in some places. I included some suggested text. Also like Tim says we need the status subresource.

**Review by rossf7 — COMMENTED** [review]
@tomahawk28 This is looking good but it still needs the godoc example of how the CR will look. You can base that on examples in the spec PR. Please also add an example for the appCatalog CR.

```
spec:
  catalog: "giant-swarm"
  app: “kubernetes-prometheus”
  release: 1.0.0
  kubeContext: “giantswarm-12345”
  namespace: “monitoring”
```

File: pkg/apis/app/v1alpha1/app_catalog_types.go [hunk 6]
```diff
@@ -46,6 +46,20 @@ func NewAppCatalogCRD() *apiextensionsv1beta1.CustomResourceDefinition {
...
+//     apiVersion: app.giantswarm.io/v1alpha1
+//     kind: AppCatalog
+//     metadata:
+//       name: "giant-swarm"
+//     spec:
+//       title: "Giant Swarm"
+//       description: “Catalog of Apps by Giant Swarm"
+//       catalogStorage:
+//         type: "helm"
+//         URL: "https://giantswarm.github.com/app-catalog/"
```
Comment (kopiczko) [inline pkg/apis/app/v1alpha1/app_catalog_types.go]:
```suggestion
//         URL: "https://github.com/giantswarm/app-catalog/"
```

File: pkg/apis/app/v1alpha1/app_catalog_types.go [hunk 7]
```diff
@@ -46,6 +46,20 @@ func NewAppCatalogCRD() *apiextensionsv1beta1.CustomResourceDefinition {
...
 
+// Example of AppCatalog CR as below.
+//
+//     apiVersion: app.giantswarm.io/v1alpha1
+//     kind: AppCatalog
+//     metadata:
+//       name: "giant-swarm"
+//     spec:
+//       title: "Giant Swarm"
+//       description: “Catalog of Apps by Giant Swarm"
```
Comment (kopiczko) [inline pkg/apis/app/v1alpha1/app_catalog_types.go]:
```suggestion
//       description: “Catalog of Apps by Giant Swarm."
```

File: pkg/apis/app/v1alpha1/app_deployment_types.go [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+	return metav1.TypeMeta{
+		APIVersion: version,
+		Kind:       kindAppDeployment,
+	}
+}
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// Example of AppDeployment CR as below.
```
Comment (kopiczko) [inline pkg/apis/app/v1alpha1/app_deployment_types.go]: To make it a valid godoc:

```suggestion
// AppDeployment CR example as below.
```

File: pkg/apis/app/v1alpha1/app_catalog_types.go [hunk 9]
```diff
@@ -46,6 +46,20 @@ func NewAppCatalogCRD() *apiextensionsv1beta1.CustomResourceDefinition {
 // +genclient
 // +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
 
+// Example of AppCatalog CR as below.
```
Comment (kopiczko) [inline pkg/apis/app/v1alpha1/app_catalog_types.go]: To make it a valid godoc:

```suggestion
// AppCatalog CR example as below.
```

File: pkg/apis/app/v1alpha1/app_deployment_types.go (hunk 2 above)
Comment (kopiczko) [reply pkg/apis/app/v1alpha1/app_deployment_types.go]: `kubeContext` is a terrible name IMO. When I read that it feels like `kubectl` context. Is is something from upstream?

@rossf7 you say

> The kubeContext will have the namespace where the operator should connect to deploy the app

Then I see:

```
	// Namespace is the namespace where the app should be deployed.
	// e.g. monitoring
	Namespace string   
```

We need to make this straight from the beggining IMO.

I understand this is just a `tillerNamespace` but we don't want to call it tiller as we want to get rid of it. What will it be replaced with then? How to call that?

**Review by kopiczko — CHANGES_REQUESTED** [review]

File: pkg/apis/app/v1alpha1/app_catalog_types.go (hunk 6 above)
Comment (tomahawk28) [reply pkg/apis/app/v1alpha1/app_catalog_types.go]: Ah, How I pass this 🤦‍♂️ Thanks

**Comment by tomahawk28** [comment]
@kopiczko From what I understand, `kubeContext` is one of the `context` from the control plane, not the namespace. is it correct? @rossf7

File: pkg/apis/app/v1alpha1/app_deployment_types.go (hunk 2 above)
Comment (rossf7) [reply pkg/apis/app/v1alpha1/app_deployment_types.go]: Damnit sry my comment didn't help here :( I'm not feeling well didn't get much sleep.

kubeContext is the name of the cluster to connect to. My understanding is this part of the UX changes. As it matches the kube context the user will have created with gsctl. However the design was done while I was away.

@kopiczko I think it makes more sense to address this in the spec PR. Please leave your feedback there if you have concerns.
https://github.com/giantswarm/giantswarm/pull/4577

On the namespace point I don't follow. Its the namespace where the app should be deployed. It isn't the namespace where tiller is installed.

Comment (kopiczko) [reply pkg/apis/app/v1alpha1/app_deployment_types.go]: Thanks for pointing me to the spec. I already left a comment there.

Re namespace: I didn't mean `namespace` -> `tillerNamespace` but `kubeContext` -> `tillerNamespace`.

**Comment by puja108** [comment]
we need to sync this with the spec again

**Comment by tomahawk28** [comment]
@puja108 I'm still waiting for the spec is merged, will do that soon after that.

**Review by rossf7 — COMMENTED** [review]
> should we move this chart CRD to application group?

Yes we want all 3 CRDs to use the new group. Can you update this PR so it matches the spec?

I hope we'll get the spec merged today by updating this PR we can prepare for that.

https://github.com/giantswarm/giantswarm/pull/4577#discussion_r242537961

**Comment by tomahawk28** [comment]
@giantswarm/team-batman @giantswarm/sig-operator I updated as an agreement that we had, let's merge this asap :)

File: pkg/apis/application/v1alpha1/app_catalog_types.go [hunk 10]
```diff
@@ -46,21 +46,35 @@ func NewAppCatalogCRD() *apiextensionsv1beta1.CustomResourceDefinition {
...
+//     apiVersion: application.giantswarm.io/v1alpha1
+//     kind: AppCatalog
+//     metadata:
+//       name: "giant-swarm"
+//     spec:
+//       title: "Giant Swarm"
+//       description: “Catalog of Apps by Giant Swarm."
+//       catalogStorage:
+//         type: "helm"
+//         URL: "https://github.com/giantswarm/app-catalog/"
```
Comment (kopiczko) [inline pkg/apis/application/v1alpha1/app_catalog_types.go]: Apparently previous version was correct. It is a github page.

```suggestion
//         URL: "https://giantswarm.github.com/app-catalog/"
```

File: pkg/apis/application/v1alpha1/app_types.go [hunk 11]
```diff
@@ -0,0 +1,127 @@
...
+//     spec:
+//       app: “kubernetes-prometheus”
+//       catalog: "giant-swarm"
+//       namespace: “monitoring”
+//       release: 1.0.0
+//
+//       kubeConfig:
+//		  secret:
+//		    name: "giantswarm-12345"
+//		    namespace: "12345"
```
Comment (kopiczko) [inline pkg/apis/application/v1alpha1/app_types.go]: Indentation is broken here.

File: pkg/apis/application/v1alpha1/app_types.go [hunk 12]
```diff
@@ -0,0 +1,127 @@
...
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// App CR example as below.
+//
+//     apiVersion: application.giantswarm.io/v1alpha1
+//     kind: App
+//     metadata:
+//       name: “My-Cool-Prometheus”
```
Comment (kopiczko) [inline pkg/apis/application/v1alpha1/app_types.go]:
```suggestion
//       name: "My-Cool-Prometheus"
```

Comment (kopiczko) [reply pkg/apis/application/v1alpha1/app_types.go]: Those quotes `”` appear in many places. Please fix them.

File: pkg/apis/application/v1alpha1/app_types.go [hunk 13]
```diff
@@ -0,0 +1,127 @@
...
+//       namespace: “monitoring”
+//       release: 1.0.0
+//
+//       kubeConfig:
+//		  secret:
+//		    name: "giantswarm-12345"
+//		    namespace: "12345"
+//
+//     status:
+//       releaseStatus: “DEPLOYED”
```
Comment (kopiczko) [inline pkg/apis/application/v1alpha1/app_types.go]: I can easily see we add more fields to the release here. I would prefer to have one level more here like below. This also would mean structs changes.

```
//       release:
//         status: “DEPLOYED”
```

File: pkg/apis/application/v1alpha1/app_types.go [hunk 14]
```diff
@@ -0,0 +1,127 @@
...
+
+// App CR example as below.
+//
+//     apiVersion: application.giantswarm.io/v1alpha1
+//     kind: App
+//     metadata:
+//       name: “My-Cool-Prometheus”
+//       namespace: “12345”
+//     spec:
+//       app: “kubernetes-prometheus”
```
Comment (kopiczko) [inline pkg/apis/application/v1alpha1/app_types.go]: I think `name` is better fit than `app` here as this is `App` kind.

Comment (tomahawk28) [reply pkg/apis/application/v1alpha1/app_types.go]: @kopiczko Can you please write that requests on spec review, We need to sync on that PR, especially about spec things at there.

File: pkg/apis/application/v1alpha1/app_types.go (hunk 13 above)
Comment (tomahawk28) [reply pkg/apis/application/v1alpha1/app_types.go]: @kopiczko this as well

**Comment by rossf7** [comment]
@kopiczko Yes it will be easier to add further feedback to the spec PR. But I like both these suggestions. I'll update the spec PR because I need to merge it soon.

File: pkg/apis/application/v1alpha1/app_types.go [hunk 15]
```diff
@@ -0,0 +1,132 @@
...
+//          namespace: "12345"
+//
+//     status:
+//       release:
+//         status: “DEPLOYED”
+//
+type App struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata"`
+	Spec              AppSpec `json:"spec"`
```
Comment (rossf7) [inline pkg/apis/application/v1alpha1/app_types.go]: Status struct is missing here.

File: pkg/apis/application/v1alpha1/app_types.go [hunk 16]
```diff
@@ -0,0 +1,132 @@
...
+	Catalog string `json:"catalog" yaml:"catalog"`
+	// Namespace is the namespace where the application should be deployed.
+	// e.g. monitoring
+	Namespace string `json:"namespace" yaml:"namespace"`
+	// Release is the version of this application which we would like to use.
+	// e.g. 1.0.0
+	Release string `json:"release" yaml:"release"`
+	// KubeConfig is the name of secret resource  which represent specific k8s cluster where the application could be deployed.
+	KubeConfig AppSpecKubeConfig `json:"kubeConfig" yaml:"kubeConfig"`
+	Status     AppSpecStatus     `json:"status" yaml:"status"`
```
Comment (rossf7) [inline pkg/apis/application/v1alpha1/app_types.go]: Status should move to the line above.

File: pkg/apis/application/v1alpha1/app_types.go [hunk 17]
```diff
@@ -0,0 +1,132 @@
...
+}
+
+type AppSpecKubeConfig struct {
+	Secret AppSpecSecret `json:"secret" yaml:"secret"`
```
Comment (rossf7) [inline pkg/apis/application/v1alpha1/app_types.go]: I think this should be `AppSpecKubeConfigSecret`.

File: pkg/apis/application/v1alpha1/app_types.go [hunk 18]
```diff
@@ -0,0 +1,132 @@
...
+}
+
+type AppSpecSecret struct {
```
Comment (rossf7) [inline pkg/apis/application/v1alpha1/app_types.go]: Same here. Should be `AppSpecKubeConfigSecret`.

File: pkg/apis/application/v1alpha1/app_types.go [hunk 19]
```diff
@@ -0,0 +1,132 @@
...
+	Name      string `json:"name" yaml:"name"`
+	Namespace string `json:"namespace" yaml:"namespace"`
+}
+
+type AppSpecStatus struct {
```
Comment (rossf7) [inline pkg/apis/application/v1alpha1/app_types.go]: Should just be `AppStatus`.

File: pkg/apis/application/v1alpha1/app_types.go [hunk 20]
```diff
@@ -0,0 +1,132 @@
...
+	Release AppSpecStatusRelease `json:"release" yaml:"release"`
+}
+
+type AppSpecStatusRelease struct {
```
Comment (rossf7) [inline pkg/apis/application/v1alpha1/app_types.go]: This struct is no longer needed.

**Review by rossf7 — COMMENTED** [review]
Some more changes needed.

File: pkg/apis/application/v1alpha1/app_types.go [hunk 21]
```diff
@@ -0,0 +1,127 @@
...
+	Namespace string `json:"namespace" yaml:"namespace"`
+	// Release is the version of this application which we would like to use.
+	// e.g. 1.0.0
+	Release string `json:"release" yaml:"release"`
+	// KubeConfig is the name of secret resource  which represent specific k8s cluster where the application could be deployed.
+	KubeConfig AppSpecKubeConfig `json:"kubeConfig" yaml:"kubeConfig"`
+}
+
+type AppSpecKubeConfig struct {
+	Secret AppSpecKubeConfigSecret `json:"secret" yaml:"secret"`
```
Comment (rossf7) [inline pkg/apis/application/v1alpha1/app_types.go]: These fields should have godoc comments.

File: pkg/apis/application/v1alpha1/app_types.go [hunk 22]
```diff
@@ -0,0 +1,127 @@
...
+}
+
+type AppSpecKubeConfigSecret struct {
+	Name      string `json:"name" yaml:"name"`
```
Comment (rossf7) [inline pkg/apis/application/v1alpha1/app_types.go]: Comments needed here too.

File: pkg/apis/application/v1alpha1/app_catalog_types.go [hunk 23]
```diff
@@ -46,21 +46,35 @@ func NewAppCatalogCRD() *apiextensionsv1beta1.CustomResourceDefinition {
...
-	// Title is the name of the app catalog for this CR
+	// Title is the name of the application catalog for this CR
 	// e.g. Catalog of Apps by Giant Swarm
 	Title       string `json:"title" yaml:"title"`
 	Description string `json:"description" yaml:"description"`
 	// CatalogStorage references a map containing values that should be
 	// applied to the appcatalog.
 	CatalogStorage AppCatalogSpecCatalogStorage `json:"catalogStorage" yaml:"catalogStorage"`
-	// LogoURL contains the links for logo image file for this app catalog
+	// LogoURL contains the links for logo image file for this application catalog
```
Comment (rossf7) [inline pkg/apis/application/v1alpha1/app_catalog_types.go]: Same here.

File: pkg/apis/application/v1alpha1/app_catalog_types.go [hunk 24]
```diff
@@ -46,21 +46,35 @@ func NewAppCatalogCRD() *apiextensionsv1beta1.CustomResourceDefinition {
...
+//
 type AppCatalog struct {
 	metav1.TypeMeta   `json:",inline"`
 	metav1.ObjectMeta `json:"metadata"`
 	Spec              AppCatalogSpec `json:"spec"`
 }
 
 type AppCatalogSpec struct {
-	// Title is the name of the app catalog for this CR
+	// Title is the name of the application catalog for this CR
```
Comment (rossf7) [inline pkg/apis/application/v1alpha1/app_catalog_types.go]: Why the change to application?

File: pkg/apis/application/v1alpha1/app_types.go [hunk 25]
```diff
@@ -0,0 +1,127 @@
...
+	// Name is the name of the application to be deployed.
+	// e.g. kubernetes-prometheus
+	Name string `json:"name" yaml:"name"`
+	// Catalog is the name of the application catalog this app belongs to.
+	// e.g. giant-swarm
+	Catalog string `json:"catalog" yaml:"catalog"`
+	// Namespace is the namespace where the application should be deployed.
+	// e.g. monitoring
+	Namespace string `json:"namespace" yaml:"namespace"`
+	// Release is the version of this application which we would like to use.
```
Comment (rossf7) [inline pkg/apis/application/v1alpha1/app_types.go]:
```suggestion
 	// Release is the version of the app that should be deployed.
```

File: pkg/apis/application/v1alpha1/app_types.go [hunk 26]
```diff
@@ -0,0 +1,127 @@
...
+	// e.g. 1.0.0
+	Release string `json:"release" yaml:"release"`
+	// KubeConfig is the name of secret resource  which represent specific k8s cluster where the application could be deployed.
```
Comment (rossf7) [inline pkg/apis/application/v1alpha1/app_types.go]:
```suggestion
 	// KubeConfig is the kubeconfig that should be used to connect to the cluster when deploying the app.
```

The current comment is specific to using a secret. In the future there might be other configuration options.

**Review by rossf7 — COMMENTED** [review]
Next round of feedback.

In the comments a lot of references to app are now application. Is that intentional? I found it easier to read before since the CRDs are app and appCatalog.

File: pkg/apis/application/v1alpha1/app_types.go (hunk 25 above)
Comment (kopiczko) [reply pkg/apis/application/v1alpha1/app_types.go]: I edited this comment to remove leading space.

File: pkg/apis/application/v1alpha1/app_types.go (hunk 21 above)
Comment (tomahawk28) [reply pkg/apis/application/v1alpha1/app_types.go]: `Secret is the reference of secret object which contains clusters info and IDs` is fine?

**Review by rossf7 — COMMENTED** [review]
@tomahawk28 This is very close. A last round of feedback and then lets get this merged.

File: pkg/apis/application/v1alpha1/app_catalog_types.go [hunk 27]
```diff
@@ -46,6 +46,20 @@ func NewAppCatalogCRD() *apiextensionsv1beta1.CustomResourceDefinition {
 // +genclient
 // +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
 
+// AppCatalog CR example as below.
```
[the remark this answers is not in the document]
Comment (tomahawk28) [reply pkg/apis/application/v1alpha1/app_catalog_types.go]: So same as app CR? okay.

[the remark this answers is not in the document]
Comment (rossf7) [reply pkg/apis/application/v1alpha1/app_catalog_types.go]: Yes please remove the CR examples from both.

File: pkg/apis/application/v1alpha1/app_types.go [hunk 28]
```diff
@@ -0,0 +1,108 @@
...
+
+type AppSpecKubeConfig struct {
+	// Secret references a secret containing the kubconfig.
+	Secret AppSpecKubeConfigSecret `json:"secret" yaml:"secret"`
+}
+
+type AppSpecKubeConfigSecret struct {
+	// Name is the name of the secret containing the kubeconfig.
+	Name      string `json:"name" yaml:"name"`
+	Namespace string `json:"namespace" yaml:"namespace"`
```
Comment (rossf7) [inline pkg/apis/application/v1alpha1/app_types.go]: Every field needs godoc. 

`// Namespace is the namespace of the secret containing the kubeconfig.`

File: pkg/apis/application/v1alpha1/app_types.go [hunk 29]
```diff
@@ -0,0 +1,108 @@
...
+}
+
+type AppStatus struct {
+	App string `json:"app" yaml:"app"`
```
Comment (rossf7) [inline pkg/apis/application/v1alpha1/app_types.go]: This should be `Status`. Godoc can be.

`// Status is the status of the deployed app.`

**Review by rossf7 — COMMENTED** [review]
Couple more changes.

**Comment by tomahawk28** [comment]
@rossf7 What about `lastUpdated` should we need it? if then, which type it should be?

**Comment by rossf7** [comment]
@tomahawk28 Lets leave `lastUpdated` for now.  We can add it later when the status resource is implemented in app-operator.

**Review by rossf7 — APPROVED** [review]
LGTM

**Comment by rossf7** [comment]
I removed Tim's review since he is rebooting for Xmas. 

@kopiczko PTAL.

**Review by kopiczko — APPROVED** [review]
LGTM

**Pull request merged**